rcl/test/rcl/test_init.cpp

```diff
@@ -98,65 +100,147 @@ struct FakeTestArgv
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+    ASSERT_FALSE(rcl_context_is_valid(&context));
+  }
+  {
+    // If argc is less than 1, argv is not null, it should be an invalid argument.
+    rcl_context_t context = rcl_get_zero_initialized_context();
+    const char * some_args[] = {"some-arg"};
+    ret = rcl_init(0, some_args, &init_options, &context);
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+    ASSERT_FALSE(rcl_context_is_valid(&context));
+  }
+  {
+    // If an invalid ROS arg is given, init should fail.
+    rcl_context_t context = rcl_get_zero_initialized_context();
+    const char * bad_remap_args[] = {
+      "some-arg", RCL_ROS_ARGS_FLAG, RCL_REMAP_FLAG, "name:="};
+    const size_t argc = sizeof(bad_remap_args) / sizeof(const char *);
+    ret = rcl_init(argc, bad_remap_args, &init_options, &context);
+    EXPECT_EQ(RCL_RET_INVALID_ROS_ARGS, ret);
+    rcl_reset_error();
+    ASSERT_FALSE(rcl_context_is_valid(&context));
+  }
+  {
+    // If an invalid enclave is given, init should fail.
+    rcl_context_t context = rcl_get_zero_initialized_context();
+    const char * bad_enclave_args[] = {
+      "some-arg", RCL_ROS_ARGS_FLAG, RCL_ENCLAVE_FLAG, "1foo"};
+    const size_t argc = sizeof(bad_enclave_args) / sizeof(const char *);
+    ret = rcl_init(argc, bad_enclave_args, &init_options, &context);
+    EXPECT_EQ(RCL_RET_ERROR, ret);
+    rcl_reset_error();
+    ASSERT_FALSE(rcl_context_is_valid(&context));
+  }
+  {
+    // If security keystore is invalid, init should fail.
+    ASSERT_TRUE(rcutils_set_env(ROS_SECURITY_ENABLE_VAR_NAME, "true"));
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+      EXPECT_TRUE(rcutils_set_env(ROS_SECURITY_ENABLE_VAR_NAME, ""));
+    });
+    ASSERT_TRUE(rcutils_set_env(ROS_SECURITY_STRATEGY_VAR_NAME, "Enforce"));
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+      EXPECT_TRUE(rcutils_set_env(ROS_SECURITY_STRATEGY_VAR_NAME, ""));
+    });
+    ASSERT_TRUE(rcutils_set_env(ROS_SECURITY_KEYSTORE_VAR_NAME, "/not/a/real/secure/root"));
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+      EXPECT_TRUE(rcutils_set_env(ROS_SECURITY_KEYSTORE_VAR_NAME, ""));
+    });
+    rcl_context_t context = rcl_get_zero_initialized_context();
+    ret = rcl_init(0, nullptr, &init_options, &context);
+    EXPECT_EQ(RCL_RET_ERROR, ret);
+    rcl_reset_error();
+    ASSERT_FALSE(rcl_context_is_valid(&context));
+  }
+  {
+    // If either the allocate or deallocate function pointers are not set,
+    // it should be invalid arg.
+    rcl_context_t context = rcl_get_zero_initialized_context();
+    init_options.impl->allocator.allocate = nullptr;
+    ret = rcl_init(0, nullptr, &init_options, &context);
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+    ASSERT_FALSE(rcl_context_is_valid(&context));
+    init_options.impl->allocator.allocate = rcl_get_default_allocator().allocate;
+    init_options.impl->allocator.deallocate = nullptr;
```

Comment:
This section can be replaced with a call to the `get_failing_allocator` defined at [`allocator_testing_utils.h`](https://github.com/ros2/rcl/blob/master/rcl/test/rcl/allocator_testing_utils.h). 

Won't block this PR, but I think we should have all the `bad allocator` utilities used for testing in just one file

Reply:
I agree. See d1963ddbed99fb924ce507329eadec6aac98289a.